### ISSUE TITLE
Save tracks without transcoding

### ISSUE THREAD

**Issue by vibee** [jrnewell/spotify-ripper#65]
Is it possible to add a feature for ripping tracks "as they are", without transcoding or decoding the audio stream to another format?

Hence, if spotify streams ogg vorbis audio with 320kb/s, I would prefer to store exactly this stream to a local file (and not losing information/quality by transcoding or massively increasing size by stroing as FLAC).
Thanks!

**Comment by jrnewell** [comment]
Unfortunately, no.  This is @Tr4sHCr4fT explanation of the situation:

> Spotify uses a proprietary implementation of OGG with added Encryption. so this script actually "plays" a track near-realtime via the official library, but pipes the raw output to an encoder instead of your sound card. FLAC can be used to store this data exactly avoiding another quality degrading transcoding pass, or as midway container for later manual conversion, but of course you will not actually get back data, which was lost in the first encoding to ogg from the audio master source by Spotify...

Both ``--wav`` and ``--flac`` options are essentially lossless encodings of the stream.  There is also the ``--pcm`` option which is basically the raw stream directed to a file (it won't play directly because there is no audio file header but can be used to transcode later).

**Comment by Tr4sHCr4fT** [comment]
i would recommend just using the ogg option with vbr q9 (default) if you want to stay close to the original. because the audio data is already optimised to beeing reproduced by vorbis decoder optimally. of course you will still degrade the quality a bit, but less than when transcoding to a different lossy codec...

**Comment by vibee** [comment]
According to http://wiki.hydrogenaud.io/index.php?title=Transcoding#Lossy-to-lossy_transcoding, transcoding from one lossy format to another significally decreases audio quality. This is true even if the same encoder / format has been used for both stages.

I am saving the files as FLAC, in order to not lose quality from the "original" ogg vorbis 320kb files. However, thanks for the quote @jrnewell . This points out why there is no such option for spotify_ripper :)

**Comment by Tr4sHCr4fT** [comment]
yes, like i said, you will loose quality when transcoding from ogg to ogg
but, it will be less of a impact than using an other lossy codec as target

why? -> each decoder algorithm decides its way which parts of audio to drop,
and using the same codec has a higher chance to produce a similar file...

**Comment by vibee** [comment]
That's not true, and is one of the big myths in digital audio processing.

The encoded ogg vorbis file is completely different to the original audio file. Ogg vorbis is tuned to be transparent / High quality on "natural audio" input, e.g. music, speech, and so on. It produces very bad quality if the input is an ogg vorbis file (see my link; 192k ogg vorbis -> 128k ogg vorbis is noticeably worse than 192k mp3 -> 128k ogg vorbis). A 320k ogg vorbis file is totally different from the original uncompressed audio file.

However, I agree that using ~320k for source and target, it might be hard to spot any audible artifacts.

**Comment by Tr4sHCr4fT** [comment]
mhh, that report is from 2005. maybe it was a bug and is now fixed?

**Comment by vibee** [comment]
It might be just the other way around nowadays. However, recommending ogg vorbis -Q 9 because source was ogg vorbis with 320k, is wrong for the reasons mentioned in my last post. The encoder throws away some information in each transcoding pass, and produces a totally different file in each pass. Ogg vorbis (and every other lossy encoder) is tuned for the first pass, where the source is an uncompressed pcm file. In the second pass, ogg vorbis removes completely different parts of the audio file than in the first pass, because the source is totally different (in a technical way, although it might sound the same to human ears). 

Besides this, an other argument might be: The implementation of spotify's vorbis encoder is different to the reference implemention which we use. We also don't know whether spotify uses -Q9 or some totally different switches.

**Comment by Tr4sHCr4fT** [comment]
Spotify states using -q 9 here:
https://support.spotify.com/de/learn-more/faq/#!/article/What-bitrate-does-Spotify-use-for-streaming

i dont know if this changes much, but the transcoding process does first produce raw pcm from ogg

**Comment by vibee** [comment]
Still, your argumentation does not hold, except you can prove your statement (ogg q9 -> ogg q9 is better than ogg q9 -> any other lossy encoder).
Of course transcoding from ogg to ogg has raw pcm in the middle, as an audio encoder takes raw pcm as input. It doesn't contradict my argumentation.
Here's one more link about generation loss within the same codec:
http://bernholdtech.blogspot.de/2013/03/Nine-different-audio-encoders-100-pass-recompression-test.html
It what you say was true (transcoding from lossy ogg vorbis to lossy ogg vorbis hardly affects audio quality because it always "removes" the same information), then the generation loss should not appear in an audible way.

**Comment by Tr4sHCr4fT** [comment]
mh, i will test it by time, but from my memory the ogg encoded file had the fullest spektogram...

**Comment by vibee** [comment]
A spectogram does not say anything about audio quality. Audio quality is not visualizable. A lossy audio file with lowpass filter of ~18khz may (and in many cases will) sound more transparent than one without.

**Comment by Tr4sHCr4fT** [comment]
what way of comparison would you suggest? 
i could not reliable abx between any of the avaiable encoder options at 256k and a flac dump, well, except for faac...

**Comment by jrnewell** [comment]
@vibee As to your original question - In case you are interested, I stumbled across a blog post http://moyix.blogspot.com/2014/07/breaking-spotify-drm-with-panda.html?m=1 a while ago that discusses using [PANDA](https://github.com/moyix/panda) to decrypt Spotify streams to pull the original unencrypted Ogg files.  So if you have some decent hacking skills it might be possible - I haven't really tried it.

Or you could look at Tidal which offers "HiFi" FLAC streaming for a premium cost.

**Comment by marianolatorre** [comment]
I was thinking to use itunes match or google music to match the songs and get fresh copies then :)
